OpenSim/Simulation/GeodesicWrap/GeodesicWrappingMath.cpp

```diff
@@ -0,0 +1,572 @@
+ImplicitSurfaceState::calcDerivative(const SimTK::Vec3& acceleration)
+const
+{
+    return {velocity, acceleration};
+}
+
+JacobiFieldScalar::Dot
+JacobiFieldScalar::calcDerivative(double gaussianCurvature)
+const
+{
+    return {derivative, -gaussianCurvature*value};
+}
+
+ImplicitGeodesicState::Dot
+ImplicitGeodesicState::calcDerivative(
+const SimTK::Vec3& acceleration,
+        double gaussianCurvature)
+const
+{
+    ImplicitGeodesicState::Dot dx;
+    const ImplicitSurfaceState q = *this;
+    // TODO magic numbers, better way to switch betweeen struct and vector?
+    dx.topRows(6) = q.calcDerivative(acceleration);
+    reinterpret_cast<JacobiFieldScalar::Dot&>(dx[6]) = a.calcDerivative(gaussianCurvature);
+    reinterpret_cast<JacobiFieldScalar::Dot&>(dx[8]) = r.calcDerivative(gaussianCurvature);
+    return dx;
+}
+
+ImplicitGeodesicState&
+ImplicitGeodesicState::
+operator+=(const ImplicitGeodesicState::Dot &derivative)
+{
+    ImplicitSurfaceState& q = *this;
+    q += derivative.topRows(6);
+    a += reinterpret_cast<const JacobiFieldScalar::Dot&>(derivative[6]);
+    r += reinterpret_cast<const JacobiFieldScalar::Dot&>(derivative[8]);
+    return *this;
+}
+
+//==============================================================================
+//                      IMPLICIT SURFACE OBJECT
+//==============================================================================
+
+GeodesicCurveState
+calcGeodesicCurveState(
+        const ImplicitSurfaceParams& s,
+        const ImplicitGeodesicState& x)
+{
+    GeodesicCurveState out;
+    out.frame = calcDarbouxFrame(s, x);
+    out.position = x.postion;
+    out.normalCurvature = calcNormalCurvature(s, x);
+    out.geodesicTorsion = calcGeodesicTorsion(s, x);
+    out.a = x.a;
+    out.r = x.r;
+    return out;
+}
+
+ImplicitSurfaceObject::ImplicitSurfaceObject(
+        ImplicitSurfaceParams&& params) :
+    _surface(std::move(params))
+{}
+
+void
+ImplicitSurfaceObject::calcInitState(
+        const SimTK::Vec3& pStart,
+        const SimTK::Vec3& pEnd)
+{
+    _startState = _surface.calcInitState(pStart, pEnd);
+    _length = _surface.calcInitLength(pStart, pEnd);
+}
+
+void
+ImplicitSurfaceObject::calcLocalGeodesic(
+        GeodesicCurveState& start,
+        GeodesicCurveState& end,
+        double& length,
+        std::vector<SimTK::Vec3>& log)
+{
+    auto geodesicEndPoints = _surface.calcLocalGeodesic(_startState, length, log);
+    start = calcGeodesicCurveState(_surface, geodesicEndPoints.first);
+    end = calcGeodesicCurveState(_surface, geodesicEndPoints.second);
+    length = _length;
+}
+
+void ImplicitSurfaceObject::applyVariation(
+        const GeodesicVariation& var)
+{
+    _surface.applyVariation(var, _startState);
+}
+
+
+//==============================================================================
+//                      WRAPOBJECT COMPONENT
+//==============================================================================
+
+WrapObject::WrapObject(
```

Comment:
As discussed in the recent meeting, we can totally ignore OpenSim's existing `WrapObject` component.